pkg/graveler/graveler.go

```diff
@@ -862,15 +879,124 @@ func generateStagingToken(repositoryID RepositoryID, branchID BranchID) StagingT
+		return nil, ErrBranchExists
+	}
+	if !errors.Is(err, ErrBranchNotFound) {
+		return nil, err
+	}
+
+	newBranch := Branch{
+		CommitID:     reference.CommitID,
+		StagingToken: generateStagingToken(repositoryID, branchID),
+	}
+	storageNamespace := repo.StorageNamespace
+	preRunID := g.hooks.NewRunID()
+	err = g.hooks.PreCreateBranchHook(ctx, HookRecord{
+		RunID:            preRunID,
+		StorageNamespace: storageNamespace,
+		EventType:        EventTypePreCreateBranch,
+		SourceRef:        ref,
+		RepositoryID:     repositoryID,
+		BranchID:         branchID,
+		CommitID:         reference.CommitID,
+	})
+	if err != nil {
+		return nil, &HookAbortError{
+			EventType: EventTypePreCreateBranch,
+			RunID:     preRunID,
+			Err:       err,
+		}
+	}
+
+	err = g.RefManager.CreateBranch(ctx, repositoryID, branchID, newBranch)
+	if err != nil {
+		return nil, fmt.Errorf("set branch '%s' to '%v': %w", branchID, newBranch, err)
+	}
+
+	postRunID := g.hooks.NewRunID()
+	g.hooks.PostCreateBranchHook(ctx, HookRecord{
+		RunID:            postRunID,
+		StorageNamespace: storageNamespace,
+		EventType:        EventTypePostCreateBranch,
+		SourceRef:        ref,
+		RepositoryID:     repositoryID,
+		BranchID:         branchID,
+		CommitID:         reference.CommitID,
+		PreRunID:         preRunID,
+	})
+
+	return &newBranch, nil
+}
+
+// checkEmptyToken Checks whether token contains etries. Returns error if List operation fails or token contains entries, False otherwise
+func (g *KVGraveler) checkEmptyToken(ctx context.Context, stagingToken StagingToken) error {
+	iter, err := g.StagingManager.List(ctx, stagingToken, 1)
+	if err != nil {
+		return err
+	}
+	defer iter.Close()
+	if iter.Next() {
+		return ErrConflictFound
+	}
+	return nil
 }
 
 func (g *KVGraveler) UpdateBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID, ref Ref) (*Branch, error) {
-	return g.db.UpdateBranch(ctx, repositoryID, branchID, ref)
+	reference, err := g.Dereference(ctx, repositoryID, ref)
+	if err != nil {
+		return nil, err
+	}
+	if reference.ResolvedBranchModifier == ResolvedBranchModifierStaging {
+		return nil, fmt.Errorf("reference '%s': %w", ref, ErrDereferenceCommitWithStaging)
+	}
+	newBranch := &Branch{}
+
+	err = g.RefManager.BranchUpdate(ctx, repositoryID, branchID, func(b *Branch) (*Branch, error) {
+		curBranch, err := g.RefManager.GetBranch(ctx, repositoryID, branchID)
+		if err != nil {
+			return nil, err
+		}
+		newBranch.CommitID = reference.CommitID
+		newBranch.StagingToken = curBranch.StagingToken
+		// validate no conflict (Check Staging Token and Sealed Tokens are empty)
+		// TODO(Guys) return error only on conflicts, currently returns error for any changes on staging
+		if err = g.checkEmptyToken(ctx, b.StagingToken); err != nil {
+			return nil, err
+		}
+		for _, st := range b.SealedTokens {
+			err = g.checkEmptyToken(ctx, st)
+			if err != nil {
+				return nil, err // First token with entries found, break
+			}
+		}
```

Comment:
There's a bug here which is really easy to overlook, so let's add a test for it.
Imagine the following scenario: 
- `StagingToken:token1`
- `SealedTokens:[token2]`
- Staging area items:
	  - 	`token1: ["key1":tombstone]`, `token2: ["key1":"some-value"]`
	  - 	Imagine that the commit doesn't have `key1`

The update will fail, but staging area is actually empty.